runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamPool.java

```diff
@@ -0,0 +1,177 @@
+   * be matched with a call of {@link WindmillStreamPool#releaseStream(WindmillStream)}. If the
+   * stream has been cached but has timed out and drained (no longer has any holds), the stream will
+   * be closed.
+   */
+  public StreamT getStream() {
+    int index = streamPool.size() == 1 ? 0 : ThreadLocalRandom.current().nextInt(streamPool.size());
+    // We will return this stream
+    StreamT resultStream;
+    StreamT closeThisStream = null;
+    try {
+      synchronized (this) {
+        WindmillStreamPool.StreamData<StreamT> existingStreamData = streamPool.get(index);
+        // There are 3 possible states that can result from fetching the stream from the cache.
+        // 1. Stream doesn't exist create and cache a new one.
+        if (existingStreamData == null) {
+          resultStream = createAndCacheStream(index).stream;
+        }
+        // 2. The stream exists, but has timed out. The timed out stream is not returned (a new one
+        // is created and returned here) and evicted from the cache if the stream has completely
+        // drained. Every call to getStream(), is matched with a call to releaseStream(), so the
+        // stream will eventually drain and be closed either there or here if we randomly fetch the
+        // stream again.
+        else if (existingStreamData.hasTimedOut(streamTimeout)) {
+          if (--existingStreamData.holds == 0) {
+            holds.remove(existingStreamData.stream);
+            closeThisStream = existingStreamData.stream;
+          }
+          // Create and cache a new stream at the closed stream's index.
+          resultStream = createAndCacheStream(index).stream;
+        } else {
+          // 3. The stream exists and is in a valid state.
+          existingStreamData.holds++;
+          resultStream = existingStreamData.stream;
+        }
+      }
+      return resultStream;
+    } finally {
+      if (closeThisStream != null) {
+        closeThisStream.close();
+      }
+    }
+  }
+
+  private synchronized WindmillStreamPool.StreamData<StreamT> createAndCacheStream(int cacheKey) {
+    WindmillStreamPool.StreamData<StreamT> newStreamData =
+        new WindmillStreamPool.StreamData<>(streamSupplier.get());
+    newStreamData.holds++;
+    streamPool.set(cacheKey, newStreamData);
+    holds.put(newStreamData.stream, newStreamData);
+    return newStreamData;
+  }
+
+  /** Releases a stream that was obtained with {@link WindmillStreamPool#getStream()}. */
+  public void releaseStream(StreamT stream) {
+    boolean closeStream = false;
+    synchronized (this) {
+      StreamData<StreamT> streamData = holds.get(stream);
+      // All streams that are created by an instance of a pool will be present.
+      if (streamData == null) {
+        throw new IllegalStateException(
+            "Attempted to release stream that does not exist in this pool. Streams not created from this pool "
+                + "cannot be released by it.");
```

Comment:
or this stream may have been released more times than acquired.

Reply:
done.